### PR TITLE
[#708] Allow command responses reply-to without device id.

### PR DESCRIPTION
**Pull request by pellmann**
- authorization to tenant is still there
- information 'device-id' encoded in 'reply-id' by 1/0 at start

### CODE REVIEW

**Comment by sophokles73** [comment]
@pellmann  some unit tests are failing ...

**Comment by pellmann** [comment]
@sophokles73 - ups.. fixed that.

File: service-base/src/main/java/org/eclipse/hono/service/command/Command.java [hunk 1]
```diff
@@ -260,11 +260,14 @@ public String getCorrelationId() {
...
-
+    static String getRequestId(final String correlationId, final String replyToId, final String deviceId) {
         final String stringOne = Optional.ofNullable(correlationId).orElse("");
-        final String stringTwo = Optional.ofNullable(replyToId).orElse("");
-        return String.format("%02x%s%s", stringOne.length(), stringOne, stringTwo);
+        String stringTwo = Optional.ofNullable(replyToId).orElse("");
+        final boolean removeDeviceFromReplyTo = stringTwo.startsWith(deviceId + "/");
+        if(removeDeviceFromReplyTo) {
+            stringTwo = stringTwo.substring(deviceId.length()+1);
+        }
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/command/Command.java]: missing whitespaces

File: service-base/src/main/java/org/eclipse/hono/service/command/Command.java [hunk 2]
```diff
@@ -123,7 +123,7 @@ public static Command from(final ProtonDelivery delivery, final Message message,
                         Objects.requireNonNull(tenantId),
                         Objects.requireNonNull(deviceId),
                         correlationId,
-                        replyTo.getResourcePath()[3]);
+                        replyTo.toString().substring(replyTo.getBasePath().length()+1));
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/command/Command.java]: I would rather add a new method (e.g. *asString(int startIdx)*) to `ResourceIdentifier`for this purpose. Or maybe even add something more specific like *getReplyToId()*. WDYT?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java [hunk 3]
```diff
@@ -485,38 +485,33 @@ protected void doStop(final Future<Void> stopFuture) {
...
             final String tenantId,
-            final String deviceId,
             final CommandResponse response) {
 
         Objects.requireNonNull(tenantId);
-        Objects.requireNonNull(deviceId);
         Objects.requireNonNull(response);
 
-        return createCommandResponseSender(tenantId, deviceId, response.getReplyToId())
+        return createCommandResponseSender(tenantId, response.getReplyToId())
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: wouldn't it make sense to add a *sendCommandResponse(CommandResponse response)* method to `CommandResponseSender`?

File: service-base/src/main/java/org/eclipse/hono/service/command/Command.java (hunk 2 above)
Comment (pellmann) [reply service-base/src/main/java/org/eclipse/hono/service/command/Command.java]: From the `ResourceIdentifier`s POV I would suggest `getWithoutBasePath()`?

Comment (pellmann) [reply service-base/src/main/java/org/eclipse/hono/service/command/Command.java]: Or `getPathWithoutBase()`?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 3 above)
Comment (pellmann) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: Since only the tenantId and replyId are needed for this I think it is clearer to stay with this signature -also only the CommandResponse would not offer the tenantId.

File: core/src/main/java/org/eclipse/hono/util/ResourceIdentifier.java [hunk 4]
```diff
@@ -230,13 +230,25 @@ public String toString() {
...
-     * 
+     *
      * @return A string consisting of the properties separated by a forward slash.
      */
     public String getBasePath() {
         return basePath;
     }
 
+    /**
+     * Gets a string representation of ths resource identifiers's parts without the base path {@see getBasePath()}.
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/ResourceIdentifier.java]: typo, should read `identifiers'`. Also, we usually put an empty line between the description and the tags. `@see` should also go down to the tags section FMPOV.

File: core/src/main/java/org/eclipse/hono/util/ResourceIdentifier.java [hunk 5]
```diff
@@ -230,13 +230,25 @@ public String toString() {
...
+     * @return A string with all parts after the base bath.
+     */
+    public String getPathWithoutBase() {
+        if (resource.length() > basePath.length()) {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/ResourceIdentifier.java]: `return createStringRepresentation(2);` should do the trick ...

File: service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java (hunk 3 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: ??? I am talking about *CommandResponseSender.sendCommandResponse()*

Comment (pellmann) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java]: Ah, I see - yes, that makes sense.

File: site/content/api/Command-And-Control-API.md [hunk 6]
```diff
@@ -26,7 +26,9 @@ The following API defines operations that can be used by *Business Applications*
...
+3. *Business Application* has established an AMQP link in role *receiver* with the source address `control/${tenant_id}/${reply-id}`. Where `${reply-id}` may be any arbitrary string chosen by the client. This link is used by the client to receive the response of the executed command. This link’s source address is also referred to as the reply-to address for the request messages. 
+
+In cases where the response should be scoped for a device (one response link per device) we suggest to add the device-id as the first part of the reply-id (`control/${tenant_id}/${device_id}/..`). This allows the adapters to minimize the information needed correlate the requests of the device with the responses, since this information (`${device_id}`) is already known and does not need to be encoded in such an request-id.
```
Comment (sophokles73) [inline site/content/api/Command-And-Control-API.md]: can we rephrase this to not use fourth person (`we suggest`)? we are not using this style anywhere in the API specs.
IMHO we also should not hint in the spec at the optimization that an adapter might or might not perform. However, it would be nice to have an example of a tenant scoped *reply-id* vs. a device scoped *reply-id* ...

File: site/diagrams/sequenceDiagrams/command_control_Success.txt [hunk 7]
```diff
@@ -2,9 +2,9 @@ participant AMQP 1.0 Network
 participant Business Application
 
 Business Application->AMQP 1.0 Network: attach("sender","control/${tenant_id}/${device_id}")
-Business Application->AMQP 1.0 Network: attach("receiver","control/${tenant_id}/${device_id}/${reply-to}")
+Business Application->AMQP 1.0 Network: attach("receiver","control/${tenant_id}/${reply-to}")
```
Comment (sophokles73) [inline site/diagrams/sequenceDiagrams/command_control_Success.txt]: what about the PNG generated from this?

**Comment by sophokles73** [comment]
@pellmann I also think that this deserves mentioning in the release notes ...

File: site/diagrams/sequenceDiagrams/command_control_Success.txt (hunk 7 above)
Comment (pellmann) [reply site/diagrams/sequenceDiagrams/command_control_Success.txt]: Good question. I changed it, but now it is gone. Thanks for spotting!
